### ISSUE TITLE
influx_inspect buildtsi is not safe to use on multiple shards concurrently

### ISSUE THREAD

**Issue by e-dard** [influxdata/influxdb#9775]
#### Situation

Multiple shards and the series file need rebuilding. As it stands there are no index directories and no `_series` directory on the database.

#### Goal

To build TSI indexes for all shards and the associated Series File as quickly as possible.

#### Process

In parallel run the following commands:

```sh
$ influx_inspect buildtsi -datadir /data/data -waldir /data/wal -database db -shard 1 &
$ influx_inspect buildtsi -datadir /data/data -waldir /data/wal -database db -shard 2 &
$ influx_inspect buildtsi -datadir /data/data -waldir /data/wal -database db -shard 3 &
$ influx_inspect buildtsi -datadir /data/data -waldir /data/wal -database db -shard 4 &
...
...
```

I would expect the indexes and the series file to be rebuilt correctly and safely. Assuming enough HW resources I would expect this to take less time than running:

```sh
$ influx_inspect buildtsi -datadir /data/data -waldir /data/wal -database db
```

#### What happens?

Almost always the result of this process is a corrupted Series File. I think this due to some missing synchronisation with respect to modifying the Series File from the `influx_inspect buildtsi` tool.

#### Requirements

  - It should be possible to build multiple shard indexes concurrently.
  - Either the tool should be safe to run in multiple processes, or there should be some sort of lock file preventing multiple invocations of the tool on the same database at the same time.
  - If the tool should only be run from one process then we should allow a user to provide multiple shards that they want the indexes rebuilt for, and have those shards rebuilt concurrently.
  - If the tool can be run from multiple processes then providing the existing `-shard` flag would be fine as that will allow users to rebuild a sub-set of shard in concurrently.
  - Users should be able to control the maximum number of OS-threads that the tool uses, e.g., with a `-p` flag or similar. It should default to the result of `runtime.GOMAXPROCS(0)`.

**Comment by reistiago** [comment]
@e-dard do you know if the same issue applies if one runs the tool concurrently (2 processes) in different databases?

**Comment by e-dard** [comment]
@reistiago I have recently resolved this via #10048. You will see very good improvements in performance now, and by default it will build indexes for all your shards concurrently.

**Comment by reistiago** [comment]
@e-dard that's great, thanks. 

Any idea on when this will be released? Either in 1.7, 1.6.x or 1.5.x (last 2 being backports I guess).  

Just asking because we need/want to migrate a few databases and need to plan when to do this, if the release happens soonish we might be able to wait a bit.

Thanks again

**Comment by e-dard** [comment]
@reistiago it will be backported to `1.6.1`, which will likely be released in the coming weeks. It will be in `1.7.x` too but that won't be for a while. It's unlikely we will backport it to `1.5.x`.

In terms of performance I'm now seeing ~1.3M series/second processed across 25 shards on my 24 core development box. The single shard performance is ~333K series/second.

I would love to know how you get on with your databases when you get around to doing that. Cheers!
